src/raftstore/store/peer.rs

```diff
@@ -994,15 +1107,108 @@ impl Peer {
                      quorum_after_change))
     }
 
+    fn transfer_leader(&mut self, peer: &metapb::Peer) {
+        info!("{} transfer leader to {:?}", self.tag, peer);
+
+        self.raft_group.transfer_leader(peer.get_id());
+    }
+
+    fn is_tranfer_leader_allowed(&self, peer: &metapb::Peer) -> bool {
+        let peer_id = peer.get_id();
+        let status = self.raft_group.status();
+
+        if !status.progress.contains_key(&peer_id) {
+            return false;
+        }
+
+        for progress in status.progress.values() {
+            if progress.state == ProgressState::Snapshot {
+                return false;
+            }
+        }
+
+        let last_index = self.get_store().last_index();
+        last_index <= status.progress[&peer_id].matched + TRANSFER_LEADER_ALLOW_LOG_LAG
+    }
+
+    fn read_local(&mut self, req: RaftCmdRequest, cb: Callback, metrics: &mut RaftProposeMetrics) {
+        metrics.local_read += 1;
+        self.handle_read(req, cb);
+    }
+
+    fn read_index(&mut self,
+                  uuid: Uuid,
+                  req: RaftCmdRequest,
+                  cb: Callback,
+                  metrics: &mut RaftProposeMetrics)
+                  -> bool {
+        metrics.read_index += 1;
+
+        let renew_lease_time = clocktime::raw_now();
+        if let Some(read) = self.pending_reads.reads.back_mut() {
+            if read.renew_lease_time + self.cfg.raft_store_max_leader_lease > renew_lease_time {
+                read.cmds.push((req, cb));
+                return false;
+            }
+        }
+
+        // Should we call pre_propose here?
+        let last_pending_read_count = self.raft_group.raft.pending_read_count();
+        let last_ready_read_count = self.raft_group.raft.ready_read_count();
+
+        self.raft_group.read_index(uuid.as_bytes().to_vec());
+
+        let pending_read_count = self.raft_group.raft.pending_read_count();
+        let ready_read_count = self.raft_group.raft.ready_read_count();
```

Comment:
No. The message may be either marked as ready immediately or pushed to pending queue.